app/controllers/application_controller.rb

```diff
@@ -5,6 +5,148 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+  get '/' do
+  	erb :index
+  end
+
+  get '/signup' do
+  	if logged_in?
+  		redirect to '/tweets'
+  	else
+  		erb :signup
+  	end
+  end
+
+  get '/login' do
+  	if logged_in?
+  		redirect to '/tweets'
+  	else
+  		erb :login
+  	end
+  end
+
+  post '/login' do
+  	user = User.find_by(username: params[:username])
+  	if user && user.authenticate(params[:password])
+  		session[:user_id] = user.id
+  		redirect to '/tweets'
+  	else
+  		redirect to '/signup'
+  	end
+  end
+
+  post '/signup' do
+  	if params[:username] == "" || params[:email] == "" || params[:password] == ""
+      redirect '/signup'
+    else
+    	user = User.new(:username => params[:username], :email => params[:email], :password => params[:password])
+    	user.save
+    	session[:user_id] = user.id
+  		redirect to '/tweets'
+  	end
+  end
+
+  get '/tweets' do
+  	if logged_in?
+  		@tweets = Tweet.all
+	  	erb :tweets
+	  else
+	  	redirect '/login'
+	  end
+  end
+
+  get '/tweets/new' do
+  	if logged_in?
+  		erb :'new_tweet'
+  	else
+  		redirect to '/login'
+  	end
+  end
+
+  post '/tweets' do
+    if params[:content] == ""
+      redirect to "/tweets/new"
+    else
+      @tweet = current_user.tweets.create(content: params[:content])
+      redirect to "/tweets/#{@tweet.id}"
+    end
+  end
+
+  get '/tweets/:id' do
+  	if logged_in?
+  		@tweet = Tweet.find_by_id(params[:id])
+  		erb :'show_tweet'
+  	else
+  		redirect to '/login'
+  	end
+  end
+
+  get "/logout" do
+  	if logged_in?
+  	  session.clear
+  	  redirect to '/login'
+  	else   
+   		redirect "/"
+   	end
+  end
+
+  get '/users/:slug' do
+  	@user = User.find_by_slug(params[:slug])
+  	erb :'user_tweets'
+  end
+
+  ###
+
+  get '/tweets/:id/edit' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet.user_id == current_user.id
+       erb :'edit_tweet'
+      else
+        redirect to '/tweets'
+      end
+    else
+      redirect to '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    if params[:content] == ""
+      redirect to "/tweets/#{params[:id]}/edit"
+    else
+      @tweet = Tweet.find_by_id(params[:id])
+      @tweet.content = params[:content]
+      @tweet.save
+      redirect to "/tweets/#{@tweet.id}"
+    end
+  end
+
+  delete '/tweets/:id/delete' do
+    if logged_in?
+      @tweet = Tweet.find_by_id(params[:id])
+      if @tweet.user_id == current_user.id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.